lib/has_content_id.rb

```diff
@@ -0,0 +1,14 @@
+module HasContentId
+  extend ActiveSupport::Concern
+
+  included do
+    before_validation :generate_content_id, on: :create
+    validates :content_id, presence: true
+  end
+
+private
```

Comment:
>You end up having to build up an 'almost' ActiveRecord object to get the callback to fire

I think you've pretty much done that already. Your test subject could be an anonymous class that just includes `ActiveModel::Model`. In fact, as a testing strategy this is probably more suitable/real-world and would allow you to use a more stateful approach. 

Something like:

```ruby
Thing = Class.new do
  include ActiveModel::Model
  include HasContentId
end
```

Removing the `on: :create` from the hook and using the block form like:

```ruby
before_validation do
  self.content_id ||= SecureRandom.uuid
end
```

... might afford you a cleaner, simpler test without the stubbing and indirection:

```ruby
thing = Thing.new
thing.validate
assert thing.content_id
```